idp-service/src/main/java/io/mosip/idp/services/AuthorizationHelperService.java

```diff
@@ -0,0 +1,228 @@
+package io.mosip.idp.services;
+
+import io.mosip.idp.core.dto.*;
+import io.mosip.idp.core.exception.IdPException;
+import io.mosip.idp.core.exception.KycAuthException;
+import io.mosip.idp.core.exception.SendOtpException;
+import io.mosip.idp.core.spi.AuthenticationWrapper;
+import io.mosip.idp.core.util.AuthenticationContextClassRefUtil;
+import io.mosip.idp.core.util.Constants;
+import io.mosip.idp.core.util.ErrorConstants;
+import io.mosip.idp.core.util.IdentityProviderUtil;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.kafka.annotation.KafkaListener;
+import org.springframework.stereotype.Component;
+import org.springframework.util.CollectionUtils;
+import org.springframework.util.StringUtils;
+import org.springframework.web.context.request.async.DeferredResult;
+
+import javax.validation.constraints.NotNull;
+import java.util.*;
+import java.util.stream.Collectors;
+
+import static io.mosip.idp.core.spi.TokenService.ACR;
+import static io.mosip.idp.core.util.Constants.*;
+import static io.mosip.idp.core.util.ErrorConstants.*;
+import static io.mosip.idp.core.util.ErrorConstants.INVALID_PERMITTED_SCOPE;
+import static io.mosip.idp.core.util.IdentityProviderUtil.ALGO_MD5;
+
+@Slf4j
+@Component
+public class AuthorizationHelperService {
+
+    private static final Map<String, DeferredResult> DEFERRED_RESULT_MAP = new HashMap<>();
+
+    @Autowired
+    private AuthenticationContextClassRefUtil authenticationContextClassRefUtil;
+
+    @Autowired
+    private AuthenticationWrapper authenticationWrapper;
+
+    @Autowired
+    private CacheUtilService cacheUtilService;
+
+    @Value("#{${mosip.idp.supported.authorize.scopes}}")
+    private List<String> authorizeScopes;
+
+
+    protected void addEntryInDeferredResultMap(String key, DeferredResult deferredResult) {
+        DEFERRED_RESULT_MAP.put(key, deferredResult);
+    }
+
+    @KafkaListener(id = "link-status-consumer", autoStartup = "true", topics = "${mosip.idp.kafka.linked-session.topic}")
+    public void consumeLinkStatus(String linkCodeHash) {
+        if(DEFERRED_RESULT_MAP.get(linkCodeHash) != null) {
+            LinkTransactionMetadata linkTransactionMetadata = cacheUtilService.getLinkedTransactionMetadata(linkCodeHash);
+            if(linkTransactionMetadata == null || linkTransactionMetadata.getLinkedTransactionId() == null) {
+                log.warn("Received link-status kafka message, but key was not found in cache / was found in invalid state. Ignoring the message :{}",
+                        linkCodeHash);
+                return;
+            }
+            DEFERRED_RESULT_MAP.get(linkCodeHash).setResult(
```

Comment:
This is going to be a large map, can we do get only once and reuse the object ? Same goes for the consumeLinkAuthCodeStatus method also